keps/sig-node/4800-cpumanager-split-l3cache/README.md

```diff
@@ -0,0 +1,914 @@
+extending the production code to implement this enhancement.
+-->
+- `/pkg/kubelet/cm/cpumanager/cpu_assignment.go`: `2024-08-26` - `93.5%`
+- `/pkg/kubelet/cm/cpumanager/cpu_manager.go`: `2024-08-26` - `74.5%`
+- `/pkg/kubelet/cm/cpumanager/policy_static.go`: `2024-08-26` - `89.5%`
+
+##### Integration tests
+
+N/A. This feature requires a e2e test for testing.
+
+##### e2e tests
+
+- e2e testing should cover the deployment of a pod that is following LLC alignment which can be verified in the CPU Manager State file
+- For e2e testing, guaranteed pods will be deployed with various CPU size requirements on our own baremetal instances across different vendor architectures and confirming the CPU assignments to uncore cache core groupings. This feature is intended for baremetal only and not cloud instances.
+- Update CI to test GCP instances of different architectures utilizing uncore cache alignment feature.
+
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature gate flag option
+- Initial e2e tests completed and enabled
+
+### Upgrade / Downgrade Strategy
+
+N/A
+
+
+### Version Skew Strategy
+
+N/A
+
+
+## Production Readiness Review Questionnaire
+
+<!--
+
+Production readiness reviews are intended to ensure that features merging into
+Kubernetes are observable, scalable and supportable; can be safely operated in
+production environments, and can be disabled or rolled back in the event they
+cause increased failures in production. See more in the PRR KEP at
+https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.
+
+The production readiness review questionnaire must be completed and approved
+for the KEP to move to `implementable` status and be included in the release.
+
+In some cases, the questions below should also have answers in `kep.yaml`. This
+is to enable automation to verify the presence of the review, and to reduce review
+burden and latency.
+
+The KEP must have a approver from the
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+-->
+
+### Feature Enablement and Rollback
+
+To enable this feature requires adding to the Kubelet yaml file for the CPUManager feature gate and add the policy option for L3 Cache alignment
+
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+For `CPUManager` it is a requirement going from `none` to `static` policy cannot be done dynamically because of the `cpu_manager_state file`. The node needs to be drained and the policy checkpoint file (`cpu_manager_state`) need to be removed before restarting Kubelet. This feature specifically relies on the `static` policy being enabled.
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: `CPUManagerPolicyOptions`
+  - Components depending on the feature gate: `kubelet`
```

Comment:
If this is not a feature that spans multiple kube components, why don't we just add a flag into kubelet configuration? Considering that most gates today don't have a clear specification regarding the component affected, users will have to change all components to add a feature gate, and that would be a huge waste and a error-prone process.

Suppose we "hide" this feature behind the `CPUManagerPolicyOptions` gate, how do we ensure that enabling this gate won't enable other features I don't want?

Reply:
thanks for catching! yes the FG specified here is obsolete, we should use `CPUManager{Alpha,Beta}PolicyOptions` indeed

Reply:
@ffromani has pointed me that there is a sig-arch approved way of adding policy options without drastically increasing feature gates.

https://github.com/kubernetes/enhancements/pull/4810/files#r1743109924

Reply:
Yes, we need alpha/beta gates for this on the stage ie CPUManagerPolicyOptionsAlpha for alpha stage. And CPUManagerPolicyOptionsBeta for beta stage.

Can you add that to the section describinng this also? I think its important enough to mention in the proposal as not every reviewer will be familar with the special CPUManagerPolicy feature gates.

Reply:
I went ahead and changed the feature, thought I did it but good catch. Adding to section describing this as well

Reply:
Are the Alpha/Beta features for CPUManager documented anywhere? This has always been my question when I'm updating the feature gates on the Kubernetes documentation website. It does matter that the reviewers know what features are behind these two (Alpha/Beta) gates, however, the more important thing is to let our users know them so that they can try them and give us feedbacks/suggestions. I know that `CPUManagerPolicyOptionsAlpha` is not safe for me to enable, but how can I be sure that enabling `CPUManagerPolicyOptionsBeta` would be okay for my workloads. I don't even know what that options are about. They are completely opaque to users. To put this in another way, these two options are only meaningful to a small group of developers. However, because they are "feature gates", they appear on the output from `kube-scheduler --help` command and `kube-proxy --help` command.

Reply:
This is what they mentioned earlier [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go#L110C1-L117C83) let me see if I can find anymore.  Maybe @ffromani has more info?

Reply:
sure, we have https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/#static-policy-options and we also have https://github.com/kubernetes/website/issues/38121 , and unfortunately I've noot been able to make progress on the latter

Reply:
If we don't have further questions on this, Ill close this out